adr/ADR-22.md

```diff
@@ -0,0 +1,184 @@
+sensitive key data might be stored.
+
+This document both unifies the location and acts as a living registry of the
+namespace within the NATS configuration directory.
+
+
+## Version History
+
+|Date      |Revision|
+|----------|--------|
+|2022-01-24|Initial proposal|
+
+
+## [Context | References | Prior Work]
+
+* [ADR-21][] on Configuration Contexts moves us closer to the model
+  described herein, but ADR-21 will need updating for the non-XDG platforms.
+* The `nsc` and `ngs` commands use other locations, scattering files around.
+* A Unix-centric model is the [XDG Base Directory Specification][xdgbase].
+
+
+## Design
+
+We mandate a standard schema for where all NATS clients should lay out their
+configuration files, and guidance for migration and handling.
+
+Our goals are:
+
+ 1. Keep files of like type together, while adhering to OS norms.
+ 2. Never break existing user configuration.
+ 3. Register which files and sub-directories exist, to act as a central
+    repository of such information, and record who can safely rely upon such
+    details.
+
+Rather than reinvent the wheel, we propose to use the layout schema of
+[Adrian-George Boston's Go implementation of XDG][adrg/xdg], which uses
+specific other locations for other operating systems, all documented in
+tables in the front README.
+
+The documentation is specific enough to allow implementations in languages
+other than Go to use the same layouts.  For Go, the `adrg/xdg` implementation
+has two dependencies, both of which we already depend upon in multiple
+`nats-io` repositories: `github.com/stretchr/testify` and `golang.org/x/sys`.
+The license is MIT and the code is stable but still seeing maintenance
+updates.  It supports, “most flavors of Unix, Windows, macOS, and Plan 9”.
+
+All configuration should live inside a directory called `nats`.  Users should
+not need to learn which other tools might drop files elsewhere and should be
+able to, minimally, discover all the configuration by files or symbolic links
+within such directories.  For configuration files, on Unix, this means that
+all per-tool configuration will be within the `~/.config/nats/` directory,
+while managed data might be within the `~/.local/share/nats/` directory.
+Environment variables and choice of OS will change that.
+
+
+### Migration
+
+Much of the existing client configuration involves authentication data such as
+NKeys or combined `.creds` files.  These are more likely than most to have
+been moved by the users to secure storage and replaced with symbolic links or
+be on encrypted mounts.  As such, automatically migrating such data is fraught
+with peril.
+
+For the time being, we instead simply mandate that files be accessible by the
+new paths; it is acceptable for client libraries to detect that the new
+location does not exist on disk but the old location does exist, and put in a
+symbolic link at the new location pointing to the old location.
```

Comment:
Granted I have like 10 apps using `~/.config` but I dont see any that make such symlinks - and I know I use several that support XDG but ignore it if old files are found.  Is there existing example of such symlinks (not against them, just curious)

Reply:
I think the difference is in whether or not the app considers the location something which has to be shared with other apps.  If it's an internal detail, then just using the old location is fine.  But if we're setting expectations for where all clients, across implementation languages, should look, then ensuring that "the required location" has a symlink to "the old location" removes the need for every client library implementation to know about historical variants to just work.

Reply:
@philpennock can you clarify:

> additionally require that `~/.config/nats` work, but in new installs it should
be a symbolic link to the `~/Library/Application Support/nats` directory.

But what are we doing for windows? Are we doing to then honor what the library says? Note that this is awesome, but it is somewhat fastidious, as we LGTM'ed to use the library which does the correct thing for in any OS (the addition of the symlink is fine I think) but the trick is it was rejected - I have since completed, and released, and this will do additional tweaks - and in the case of windows will put things completely elsewhere.